list.go

```diff
@@ -0,0 +1,92 @@
+package pack
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"reflect"
+
+	"github.com/renproject/surge"
+)
+
+type List struct {
+	T     Type
+	Elems []Value
+}
+
+func NewList(vs ...Value) (List, error) {
+	if len(vs) == 0 {
+		return List{}, nil
+	}
+
+	elems := make([]Value, len(vs))
+	var t Type
+	for i := range elems {
+		// Verify the list elements have a consistent type.
+		if t == nil {
+			t = vs[i].Type()
+		} else if vs[i].Type() != t {
+			return List{}, fmt.Errorf("inconsistent list type: expected %v, got %v", t, elems[i].Type())
+		}
+		elems[i] = vs[i]
+	}
+	return List{
+		T:     t,
+		Elems: elems,
+	}, nil
+}
+
+// Type returns the list type.
+func (v List) Type() Type {
+	return typeList{
+		Type: v.T,
+	}
+}
+
+// SizeHint returns the number of bytes required to represent the list in
+// binary.
+func (v List) SizeHint() int {
+	return surge.SizeHint(v.Elems)
+}
+
+// Marshal the list into binary.
+func (v List) Marshal(buf []byte, rem int) ([]byte, int, error) {
+	return surge.Marshal(v.Elems, buf, rem)
+}
+
+// MarshalJSON marshals the list to JSON.
+func (v List) MarshalJSON() ([]byte, error) {
+	raw := []interface{}{}
+	for _, elem := range v.Elems {
+		rawField, err := json.Marshal(elem)
```

Comment:
```suggestion
		rawField, err := elem.MarshalJSON()
```
Slightly more performant.